File: PC/src/test/java/network/PacketHeaderTest.java
package network;

import org.junit.jupiter.api.Test;
import protocol.PacketHeader;

import java.io.*;

import static org.junit.jupiter.api.Assertions.*;

class PacketHeaderTest {

    @Test
    void headerLengthConstantShouldBe29() {
        assertEquals(29, PacketHeader.HEADER_LENGTH);
    }

    @Test
    void defaultConstructorShouldSetZeroes() {
        PacketHeader h = new PacketHeader();
        assertEquals(0, h.getType());
        assertEquals(0, h.getNameLength());
        assertEquals(0, h.getBodyLength());
        assertEquals(0, h.getOffset());
        assertEquals(0, h.getTotalLength());
    }

    @Test
    void parameterizedConstructorShouldSetFields() {
        PacketHeader h = new PacketHeader((byte) 4, (short) 10, 2048L, 512L, 8192L);
        assertEquals(4, h.getType());
        assertEquals(10, h.getNameLength());
        assertEquals(2048, h.getBodyLength());
        assertEquals(512, h.getOffset());
        assertEquals(8192, h.getTotalLength());
    }

    @Test
    void writeToAndReadFromShouldRoundtrip() throws IOException {
        PacketHeader original = new PacketHeader((byte) 2, (short) 7, 123456L, 0L, 999999L);

        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        original.writeTo(new DataOutputStream(baos));

        PacketHeader parsed = PacketHeader.readFrom(new DataInputStream(new ByteArrayInputStream(baos.toByteArray())));

        assertEquals(original.getType(), parsed.getType());
        assertEquals(original.getNameLength(), parsed.getNameLength());
        assertEquals(original.getBodyLength(), parsed.getBodyLength());
        assertEquals(original.getOffset(), parsed.getOffset());
        assertEquals(original.getTotalLength(), parsed.getTotalLength());
    }

    @Test
    void readFromShouldThrowOnMagicMismatch() throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        dos.writeShort(0xBEEF);
        dos.writeByte(1);
        dos.writeShort(0);
        dos.writeLong(0);
        dos.writeLong(0);
        dos.writeLong(0);

        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(baos.toByteArray()));
        assertThrows(IOException.class, () -> PacketHeader.readFrom(dis));
    }

    @Test
    void settersShouldUpdateFields() {
        PacketHeader h = new PacketHeader();
        h.setType((byte) 3);
        h.setNameLength((short) 5);
        h.setBodyLength(100L);
        h.setOffset(50L);
        h.setTotalLength(200L);

        assertEquals(3, h.getType());
        assertEquals(5, h.getNameLength());
        assertEquals(100, h.getBodyLength());
        assertEquals(50, h.getOffset());
        assertEquals(200, h.getTotalLength());
    }

    @Test
    void writeToShouldProduceExactly29Bytes() throws IOException {
        PacketHeader h = new PacketHeader((byte) 1, (short) 4, 256L, 0L, 256L);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        h.writeTo(new DataOutputStream(baos));
        assertEquals(29, baos.size());
    }

    @Test
    void toStringShouldContainFields() {
        PacketHeader h = new PacketHeader((byte) 1, (short) 0, 10L, 0L, 10L);
        String s = h.toString();
        assertTrue(s.contains("type=1"));
        assertTrue(s.contains("bodyLength=10"));
    }
}

File: DocumentSend/app/src/test/java/com/example/documentsend/network/PacketHeaderTest.kt
package com.example.documentsend.network

import org.junit.Assert.*
import org.junit.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

class PacketHeaderTest {

    @Test
    fun `MAGIC_NUMBER should be 0xCAFE`() {
        assertEquals(0xCAFE.toShort(), PacketHeader.MAGIC_NUMBER)
    }

    @Test
    fun `HEADER_LENGTH should be 29`() {
        assertEquals(29, PacketHeader.HEADER_LENGTH)
    }

    @Test
    fun `writeTo and readFrom should roundtrip correctly`() {
        val header = PacketHeader(
            type = PacketType.FILE.value,
            nameLength = 10,
            bodyLength = 1024L,
            offset = 0L,
            totalLength = 1024L
        )

        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        header.writeTo(dos)

        val bais = ByteArrayInputStream(baos.toByteArray())
        val dis = DataInputStream(bais)
        val readHeader = PacketHeader.readFrom(dis)

        assertEquals(header.type, readHeader.type)
        assertEquals(header.nameLength, readHeader.nameLength)
        assertEquals(header.bodyLength, readHeader.bodyLength)
        assertEquals(header.offset, readHeader.offset)
        assertEquals(header.totalLength, readHeader.totalLength)
    }

    @Test
    fun `writeTo should write exactly HEADER_LENGTH bytes`() {
        val header = PacketHeader(
            type = PacketType.TEXT.value,
            nameLength = 0,
            bodyLength = 100L,
            offset = 0L,
            totalLength = 100L
        )

        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        header.writeTo(dos)

        assertEquals(PacketHeader.HEADER_LENGTH, baos.size())
    }

    @Test
    fun `readFrom should throw IOException for wrong magic number`() {
        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        dos.writeShort(0xFFFF.toInt()) // wrong magic number
        dos.writeByte(1)
        dos.writeShort(0)
        dos.writeLong(100L)
        dos.writeLong(0L)
        dos.writeLong(100L)

        val bais = ByteArrayInputStream(baos.toByteArray())
        val dis = DataInputStream(bais)

        assertThrows(IOException::class.java) {
            PacketHeader.readFrom(dis)
        }
    }

    @Test
    fun `header with offset should roundtrip correctly`() {
        val header = PacketHeader(
            type = PacketType.FILE.value,
            nameLength = 5,
            bodyLength = 2048L,
            offset = 512L,
            totalLength = 2048L
        )

        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        header.writeTo(dos)

        val bais = ByteArrayInputStream(baos.toByteArray())
        val dis = DataInputStream(bais)
        val readHeader = PacketHeader.readFrom(dis)

        assertEquals(512L, readHeader.offset)
        assertEquals(2048L, readHeader.totalLength)
    }

    @Test
    fun `header with zero nameLength should roundtrip correctly`() {
        val header = PacketHeader(
            type = PacketType.TEXT.value,
            nameLength = 0,
            bodyLength = 50L,
            offset = 0L,
            totalLength = 50L
        )

        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        header.writeTo(dos)

        val bais = ByteArrayInputStream(baos.toByteArray())
        val dis = DataInputStream(bais)
        val readHeader = PacketHeader.readFrom(dis)

        assertEquals(0.toShort(), readHeader.nameLength)
    }

    @Test
    fun `header with large bodyLength should roundtrip correctly`() {
        val largeSize = 5L * 1024 * 1024 * 1024 // 5GB
        val header = PacketHeader(
            type = PacketType.FILE.value,
            nameLength = 20,
            bodyLength = largeSize,
            offset = 0L,
            totalLength = largeSize
        )

        val baos = ByteArrayOutputStream()
        val dos = DataOutputStream(baos)
        header.writeTo(dos)

        val bais = ByteArrayInputStream(baos.toByteArray())
        val dis = DataInputStream(bais)
        val readHeader = PacketHeader.readFrom(dis)

        assertEquals(largeSize, readHeader.bodyLength)
    }

    @Test
    fun `data class equals should work correctly`() {
        val h1 = PacketHeader(type = 1, nameLength = 10, bodyLength = 100L, offset = 0L, totalLength = 100L)
        val h2 = PacketHeader(type = 1, nameLength = 10, bodyLength = 100L, offset = 0L, totalLength = 100L)
        val h3 = PacketHeader(type = 2, nameLength = 10, bodyLength = 100L, offset = 0L, totalLength = 100L)

        assertEquals(h1, h2)
        assertNotEquals(h1, h3)
    }
}
